haystack/components/rankers/transformers_similarity.py

```diff
@@ -127,6 +127,29 @@ def to_dict(self) -> Dict[str, Any]:
             model_kwargs=self.model_kwargs,
         )
 
+        model_kwargs = serialization_dict["init_parameters"]["model_kwargs"]
+        for key, value in model_kwargs.items():
+            # convert torch.dtype to string for serialization
+            if key in ["torch_dtype", "bnb_4bit_compute_dtype"] and isinstance(value, torch.dtype):
```

Comment:
I'm not entirely sure if `bnb_4bit_compute_dtype` lives directly in `model_kwargs`. I think it is usually specified by passing a Quantization Config object (see docs [here](https://huggingface.co/docs/transformers/main_classes/quantization#transformers.BitsAndBytesConfig)) using the keyword `quantization_config` (see docs [here](https://huggingface.co/docs/transformers/v4.36.1/en/main_classes/model#transformers.PreTrainedModel.from_pretrained.quantization_config))

Reply:
So I think to specify 4bit loading with this option we would use something like 
```python
from haystack.components.rankers import TransformersSimilarityRanker
from haystack import Document

ranker = TransformersSimilarityRanker(
  model_name_or_path="cross-encoder/ms-marco-MiniLM-L-6-v2",
  top_k=2,
  model_kwargs={"quantization_config": BitsAndBytesConfig(load_in_4bit=True, bnb_4bit_compute_dtype=torch.float16)}
)
ranker.warm_up()
```

Reply:
When working on the HFLocalGenerator, I have verified that putting the quantization configuration in `model_kwargs` properly works.

This works:
```python
model_kwargs = {
        "load_in_4bit": True,
        "bnb_4bit_use_double_quant": True,
        "bnb_4bit_quant_type": "nf4",
        "bnb_4bit_compute_dtype": torch.bfloat16,
    }

model = AutoModelForSequenceClassification.from_pretrained("cross-encoder/ms-marco-MiniLM-L-6-v2", **model_kwargs)

print(model.__dict__)
#   ...
#   "quantization_config": {
#     "bnb_4bit_compute_dtype": "bfloat16",
#     "bnb_4bit_quant_type": "nf4",
#     "bnb_4bit_use_double_quant": true,
#     "llm_int8_enable_fp32_cpu_offload": false,
#     "llm_int8_has_fp16_weight": false,
#     "llm_int8_skip_modules": null,
#     "llm_int8_threshold": 6.0,
#     "load_in_4bit": true,
#     "load_in_8bit": false,
#     "quant_method": "bitsandbytes"
#   },
#   ...
```

It does not work out of the box in the `TransformersSimilarityRanker` because `device_map` is not yet supported (I have seen you are working on that in #6679 🙏).

Reply:
Nice! Do we also want to support serialization of the HFLocalGenerator if the user passes these types into the `quantization_config` into `model_kwargs` so 
```python
model_kwargs = {
        "quantization_config": {
            "load_in_4bit": True,
            "bnb_4bit_use_double_quant": True,
            "bnb_4bit_quant_type": "nf4",
           "bnb_4bit_compute_dtype": torch.bfloat16,
    }, 
}
```
I think that is currently missed with the current to_dict method

Reply:
Done. thanks!